### PR TITLE
ZoomRoom plugin - Convert for JWT to OAuth

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

Zoom is deprecating JWT authentication to their api in June 2023. These changes convert the plugin interface from using JWT to Zoom's Server-to-Server OAuth authentication.  

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Update plugin to use Server-to-Server OAuth authentication instead of JWT. 

---------

##### Who reported, requested, or paid for the change?

FPG

---------

##### Does this update or add options to the block UI?

no

---------

##### What files does it affect?

* ZoomDotNetFramework/Responses/OauthErrorResponse.cs
* ZoomDotNetFramework/Responses/OauthTokenResponse.cs
* ZoomDotNetFramework/ZoomApi.cs
* ZoomDotNetFramework/ZoomDotNetFramework.csproj
* rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs
* rocks.kfs.Zoom/Settings.cs
* rocks.kfs.Zoom/Zoom.cs
* rocks.kfs.Zoom/rocks.kfs.Zoom.csproj

---------

##### Is it a breaking change for other versions/clients?

https://github.com/KingdomFirst/RockBlocks/pull/129 is dependent on this change

### CODE REVIEW

File: ZoomDotNetFramework/Responses/OauthTokenResponse.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿// <copyright>
+// Copyright 2021 by Kingdom First Solutions
```
Comment (guymaness2) [inline ZoomDotNetFramework/Responses/OauthTokenResponse.cs:2]:
```suggestion
// Copyright 2023 by Kingdom First Solutions
```

File: ZoomDotNetFramework/ZoomApi.cs [hunk 2]
```diff
@@ -46,7 +83,7 @@ public T Execute<T>( RestRequest request ) where T : new()
                 BaseUrl = new System.Uri( BaseUrl )
             };
             request.AddHeader( "content-type", "application/json" );
-            request.AddHeader( "authorization", "Bearer " + _jwtToken ); // used on every request
+            request.AddHeader( "authorization", "Bearer " + _oAuthToken ); // used on every request
```
Comment (guymaness2) [inline ZoomDotNetFramework/ZoomApi.cs:86]:
```suggestion
            request.AddHeader( "authorization", "Bearer " + _oAuthToken );
```

File: rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{
+    [MigrationNumber( 9, "1.12.4" )]
+    public partial class UpdateApiAttributesForOauth : Migration
+    {
+        /// <summary>
+        /// Operations to be performed during the upgrade process.
+        /// </summary>
+        public override void Up()
+        {
+            // Create new Zoom OAuth App attributes
```
Comment (guymaness2) [inline rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs:30]: Delete old attributes as well

File: rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            RockMigrationHelper.UpdateEntityAttribute( "rocks.kfs.Zoom.ZoomApp", Rock.SystemGuid.FieldType.ENCRYPTED_TEXT, "", "", "Zoom OAuth App Account Id", "The Application Id for the Zoom Marketplace Server-to-Server OAuth app to use for KFS Zoom integration elements.", 0, "", "803991F1-9FA1-4AF7-94A7-D1748CBA764F", "rocks.kfs.ZoomAppAccountId" );
+            RockMigrationHelper.UpdateEntityAttribute( "rocks.kfs.Zoom.ZoomApp", Rock.SystemGuid.FieldType.ENCRYPTED_TEXT, "", "", "Zoom OAuth App Client Id", "The Client Id for the Zoom Marketplace Server-to-Server OAuth app to use for KFS Zoom integration elements.", 0, "", "D66B02B7-8298-4371-B1FB-B02AC5C552D0", "rocks.kfs.ZoomAppClientId" );
+            RockMigrationHelper.UpdateEntityAttribute( "rocks.kfs.Zoom.ZoomApp", Rock.SystemGuid.FieldType.ENCRYPTED_TEXT, "", "", "Zoom OAuth App Client Secret", "The Client Secret for the Zoom Marketplace Server-to-Server OAuth app to use for KFS Zoom integration elements.", 0, "", "9A1CC332-E7CE-4A0A-860E-513A73E8C2F8", "rocks.kfs.ZoomAppClientSecret" );
+        }
```
Comment (guymaness2) [inline rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs:34]: Looks like Room Reservation ReservationType model may have changed. Client having CreateDB migration errors. Need to look into issue and fix.

File: rocks.kfs.Zoom/Zoom.cs [hunk 5]
```diff
@@ -69,13 +69,14 @@ public static void SyncZoomRoomDT( RockContext rockContext = null, bool enableLo
...
 
             // Delete DefinedValue for any rooms that no longer exist
-            var zoomRoomsToDelete = zoomRoomDT.DefinedValues.Where( v => !zrList.Select( zr => zr.Zr_Id ).Contains( v.Value ) );
-            foreach ( var zrDV in zoomRoomsToDelete )
-            {
-                zoomRoomDT.DefinedValues.Remove( zrDV );
-            }
+            var zrIds = new List<string>();
+            zrIds.AddRange( zrList.Select( zr => zr.Zr_Id ).ToList() );
+            var zoomRoomsToDelete = zoomRoomDV.Where( v => !zrIds.Contains( v.Value ) );
```
Comment (guymaness2) [inline rocks.kfs.Zoom/Zoom.cs:78]:
```suggestion
            var zrIds = zrList.Select( zr => zr.Zr_Id ).ToList();
            var zoomRoomsToDelete = zoomRoomDV.Where( v => !zrIds.Contains( v.Value ) );
```

**Review by nateh777 — CHANGES_REQUESTED** [review]
Changes in comments from Guy.

File: rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs (hunk 4 above)
Comment (guymaness2) [reply rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs:34]: Fixed

File: rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs (hunk 3 above)
Comment (guymaness2) [reply rocks.kfs.Zoom/Migrations/009_UpdateApiAttributesForOauth.cs:30]: done

**Comment by nateh777** [comment]
Let's remove libs/com.centralaz.roommanagement.dll, not necessary if we do not reference it in any of our projects.

File: ZoomDotNetFramework/Enums/RoomStatus.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿// <copyright>
+// Copyright 2021 by Kingdom First Solutions
```
Comment (nateh777) [inline ZoomDotNetFramework/Enums/RoomStatus.cs:2]:
```suggestion
// Copyright 2023 by Kingdom First Solutions
```

File: ZoomDotNetFramework/Responses/ListRoomsResponse.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+﻿// <copyright>
+// Copyright 2021 by Kingdom First Solutions
```
Comment (nateh777) [inline ZoomDotNetFramework/Responses/ListRoomsResponse.cs:2]:
```suggestion
// Copyright 2023 by Kingdom First Solutions
```

File: ZoomDotNetFramework/Responses/ListRoomsResponse.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+//
+using System.Collections.Generic;
+using ZoomDotNetFramework.Entities;
+
```
Comment (nateh777) [inline ZoomDotNetFramework/Responses/ListRoomsResponse.cs:19]:
```suggestion
```

File: ZoomDotNetFramework/ZoomApi.cs [hunk 9]
```diff
@@ -172,27 +207,52 @@ public bool DeleteMeeting( long meetingId, string occurrenceId = null, bool noti
...
+                request.AddParameter( "query_name", roomName );
+            }
+            if ( unassignedOnly )
+            {
+                request.AddParameter( "unassigned_rooms", bool.TrueString );
+            }
+            if ( pageSize.HasValue && pageSize.Value > 0 )
+            {
+                var pageSizeInt = pageSize.Value < 300 ? pageSize.Value : 300;  // Max for api is 300
+                request.AddParameter( "unassigned_rooms", pageSizeInt );
```
Comment (nateh777) [inline ZoomDotNetFramework/ZoomApi.cs:241]:
```suggestion
                request.AddParameter( "page_size", pageSizeInt );
```

**Review by nateh777 — DISMISSED** [review]

**Review by nateh777 — APPROVED** [review]

**Pull request merged**